src/Sylius/Component/Promotion/Generator/CouponGenerator.php

```diff
@@ -27,58 +30,77 @@ class CouponGenerator implements CouponGeneratorInterface
     protected $couponFactory;
 
     /**
-     * @var RepositoryInterface
+     * @var CouponRepositoryInterface
      */
-    protected $repository;
+    protected $couponRepository;
 
     /**
-     * @var EntityManagerInterface
+     * @var ObjectManager
      */
-    protected $manager;
+    protected $objectManager;
+
+    /**
+     * @var GenerationPolicyInterface
+     */
+    protected $generationPolicy;
+
+    /**
+     * @var CouponInterface[]
+     */
+    protected $generatedCoupons = [];
 
     /**
      * @param FactoryInterface $couponFactory
-     * @param RepositoryInterface $repository
-     * @param EntityManagerInterface $manager
+     * @param CouponRepositoryInterface $couponRepository
+     * @param ObjectManager $objectManager
+     * @param GenerationPolicyInterface $generationPolicy
      */
-    public function __construct(FactoryInterface $couponFactory, RepositoryInterface $repository, EntityManagerInterface $manager)
-    {
+    public function __construct(
+        FactoryInterface $couponFactory,
+        CouponRepositoryInterface $couponRepository,
+        ObjectManager $objectManager,
+        GenerationPolicyInterface $generationPolicy
+    ) {
         $this->couponFactory = $couponFactory;
-        $this->repository = $repository;
-        $this->manager = $manager;
+        $this->couponRepository = $couponRepository;
+        $this->objectManager = $objectManager;
+        $this->generationPolicy = $generationPolicy;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function generate(PromotionInterface $promotion, Instruction $instruction)
+    public function generate(PromotionInterface $promotion, InstructionInterface $instruction)
     {
-        $generatedCoupons = [];
+        $this->assertGenerationIsPossible($instruction);
         for ($i = 0, $amount = $instruction->getAmount(); $i < $amount; ++$i) {
             $coupon = $this->couponFactory->createNew();
             $coupon->setPromotion($promotion);
-            $coupon->setCode($this->generateUniqueCode());
+            $coupon->setCode($this->generateUniqueCode($instruction->getCodeLength()));
             $coupon->setUsageLimit($instruction->getUsageLimit());
             $coupon->setExpiresAt($instruction->getExpiresAt());
 
-            $generatedCoupons[] = $coupon;
+            $this->generatedCoupons[] = $coupon;
```

Comment:
Event `$this->generatedCoupons[$code] = true`, so we do not have to store the entities.